File: android/src/main/java/com/capacitorjs/community/advertisingid/AdvertisingIdPlugin.kt
package com.capacitorjs.community.advertisingid

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "AdvertisingId")
public class AdvertisingIdPlugin : Plugin() {
    private val implementation = AdvertisingId()

    @PluginMethod
    public fun requestTracking(call: PluginCall) {
        val status = implementation.getTrackingStatus(context)

        val output = JSObject()
        output.put("value", status)
        call.resolve(output)
    }

    @PluginMethod
    public fun getAdvertisingId(call: PluginCall) {
        val id = implementation.getTrackingId(context)
        val status = implementation.getTrackingStatus(context)

        val output = JSObject()
        output.put("id", id)
        output.put("status", status)
        call.resolve(output)
    }

    @PluginMethod
    public fun getAdvertisingStatus(call: PluginCall) {
        val status = implementation.getTrackingStatus(context)

        val output = JSObject()
        output.put("status", status)
        call.resolve(output)
    }
}

File: ios/Sources/AdvertisingIdPlugin/AdvertisingIdPlugin.swift
import Foundation
import Capacitor

@objc(AdvertisingIdPlugin)
public class AdvertisingIdPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "AdvertisingIdPlugin"
    public let jsName = "AdvertisingId"
    public let pluginMethods: [CAPPluginMethod] = [
        .async("requestTracking", AdvertisingIdPlugin.requestTracking),
        .promise("getAdvertisingId", AdvertisingIdPlugin.getAdvertisingId),
        .promise("getAdvertisingStatus", AdvertisingIdPlugin.getAdvertisingStatus)
    ]
    private let implementation = AdvertisingId()

    /// Presents the App Tracking Transparency prompt, from the main actor, and returns the answer.
    @MainActor
    func requestTracking(_ call: CAPPluginCall) async -> JSObject {
        let status = await implementation.requestTracking()
        return ["value": status.name()]
    }

    func getAdvertisingId(_ call: CAPPluginCall) {
        let id = implementation.getAdvertisingId()
        let status = implementation.getAdvertisingStatus()
        call.resolve([
            "id": id,
            "status": status.name()
        ])
    }

    func getAdvertisingStatus(_ call: CAPPluginCall) {
        let status = implementation.getAdvertisingStatus()
        call.resolve([
            "status": status.name()
        ])
    }
}
